Adds Tracks Analytics on backend

mehmoodak opened the pull request:
### Description

Adds Tracks Analytics by using the Telemetry library available in MU Plugins which will only work if site is hosted on VIP Platform or if the analytics is enabled via filter.

### How to test?

1. Create a site using `vip dev-env` which will have the MU Plugins.
1. Move `remote-data-blocks` plugin to the `plugins` folder.
1. Add filter in `plugin-loader.php` to enable tracking.

    ```php
    add_filter( 'remote_data_blocks_enable_tracks_analytics', '__return_true' );
    ```
1. Comment [these lines](https://github.com/Automattic/remote-data-blocks/blob/78c5cd5a7fdb345b5825932551ca8fc738ac6927/inc/Analytics/TracksAnalytics.php#L32C1-L35C4) to allow tracking on `local` env.
1. Trigger track events, verify that the events are logged on Live View without any error.

mehmoodak committed:
➕ ADDS: function to check if analytics is enabled

mehmoodak committed:
➕ ADDS: track_event function

mehmoodak committed:
➕ ADDS: events for activation and deactivation

mehmoodak committed:
➕ ADDS: remotedatablocks_data_sources_stats event

mehmoodak committed:
➕ ADDS: remotedatablocks_data_sources events

mehmoodak committed:
🐛 FIX: track prefix

mehmoodak committed:
🔨 REFACTOR: change event name

mehmoodak committed:
💫 UPDATE: remotedatablocks_view_data_sources event to only track once a day

mehmoodak committed:
💫 UPDATE: event name

mehmoodak committed:
Merge branch 'trunk' into cafe-942/add-tracks-analytics

mehmoodak committed:
➕ ADDS: remotedatablocks_usage_stats event

mehmoodak committed:
🔨 REFACTOR: move hooks tracking to Analytics class

mehmoodak committed:
🔨 REFACTOR: class name from Tracks to TracksAnalytics

mehmoodak committed:
🔨 REFACTOR: remove is_enabled function from TracksAnalytics

mehmoodak committed:
Merge branch 'trunk' into cafe-942/add-tracks-analytics

mehmoodak committed:
➕ ADDS: tests for get_service method

mehmoodak committed:
➕ ADDS: tests for get_data_source method

mehmoodak committed:
move tracks code to functions for easy mocking

mehmoodak committed:
update property $tracks to $instance

mehmoodak committed:
update TracksAnalytics to remove static properties and functions

mehmoodak committed:
➕ ADDS: tests for TracksAnalyticsTest

mehmoodak committed:
➕ ADDS: few tests for TracksAnalytics

mehmoodak committed:
🐛 FIX: tests

mehmoodak committed:
🐛 FIX: psalm error

mehmoodak committed:
Merge branch 'trunk' into cafe-942/add-tracks-analytics

mehmoodak committed:
➕ ADDS: tests for TracksAnalyticsTest

mehmoodak commented on tests/inc/stubs.php:
```diff
@@ -17,7 +17,15 @@ function esc_html( string $text ): string {
 }
 
-function register_block_pattern( string $name, array $options ): void {
+function register_block_pattern( string $_name, array $_options ): void {
```
psalm was complaining about unused parameters so prefixed them with `_`

mehmoodak commented on remote-data-blocks.php:
```diff
@@ -34,16 +36,19 @@
 
-// Example API
+// Analytics.
+$GLOBALS['rdb_tracks'] = new TracksAnalytics();
```
- Avoided static methods on class because of issues during mocking.
- Used global here to avoid instantiation before recording every event.

chriszarate replied:
We should solve these issues instead of introducing a global

mehmoodak replied:
Issue resolved. Thanks for your help.

mehmoodak committed:
➖ REMOVE: full stop from comments

mehmoodak committed:
👌 IMPROVE: comments

chriszarate commented on inc/Config/DataSource/HttpDataSource.php:
```diff
@@ -36,10 +36,17 @@ public function get_image_url(): ?string {
+	 * Get the service name.
+	 */
+	public function get_service(): ?string {
+		return isset( $this->config['service'] ) ? $this->config['service'] : null;
```
fyi PHP has nullish coalescing since 7.0
```suggestion
		return $this->config['service'] ?? null;
```

chriszarate commented on psalm.xml:
```diff
@@ -20,7 +20,10 @@
     <stubs>
+        <file name="vendor/php-stubs/wordpress-globals/wordpress-globals.php"/>
         <file name="vendor/php-stubs/wordpress-stubs/wordpress-stubs.php"/>
+        <file name="vendor/php-stubs/wordpress-stubs/wordpress-stubs.php"/>
```
duplicate
```suggestion
```

mehmoodak committed:
💫 UPDATE: methods to static

mehmoodak committed:
Remove duplicate from psalm.xml

Co-authored-by: Chris Zarate <chris.zarate@automattic.com>

mehmoodak committed:
Use nullish coalescing operator

Co-authored-by: Chris Zarate <chris.zarate@automattic.com>

chriszarate committed:
Add apply filters global

mehmoodak committed:
Merge branch 'trunk' into cafe-942/add-tracks-analytics

mehmoodak committed:
💫 UPDATE: tests to work properly with static methods

mehmoodak committed:
Merge branch 'trunk' into cafe-942/add-tracks-analytics

mehmoodak committed:
➕ ADDS: avoid tracking events on local env

mehmoodak committed:
Merge branch 'trunk' into cafe-942/add-tracks-analytics

mehmoodak committed:
💫 UPDATE: variable name

mehmoodak committed:
👌 IMPROVE: comments

mehmoodak committed:
👌 IMPROVE: naming

mehmoodak committed:
👌 IMPROVE: additional track props on DataSourceController

mehmoodak committed:
🐛 FIX: tracking of plugin activation and deactivation

mehmoodak committed:
💫 UPDATE: remove props that are being sent by MU Plugins

mehmoodak committed:
Merge branch 'trunk' into cafe-942/add-tracks-analytics

mehmoodak committed:
Refactor EnvironmentConfig and TracksAnalytics classes

These changes improve the overall structure and maintainability of the codebase.

mehmoodak committed:
🐛 FIX: undefined wp function errors

mehmoodak committed:
🐛 FIX: ignore psalm lint error

chriszarate commented on inc/Analytics/EnvironmentConfig.php:
```diff
@@ -0,0 +1,81 @@
+	private array $tracks_core_props = [];
+
+	public function __construct() {
+		if ( function_exists( 'Automattic\VIP\Telemetry\Tracks\get_tracks_core_properties' ) ) {
```
Does this work? I would expect that backslashes need to be escaped
```suggestion
		if ( function_exists( 'Automattic\\VIP\\Telemetry\\Tracks\\get_tracks_core_properties' ) ) {
```

mehmoodak replied:
Yes it works, seems like we don't need escaping in this case (using same syntax in MU Plugins also).

chriszarate commented on inc/Analytics/EnvironmentConfig.php:
```diff
@@ -0,0 +1,81 @@
+	}
+
+	public function get_tracks_lib_class(): ?string {
+		if ( ! class_exists( 'Automattic\VIP\Telemetry\Tracks' ) ) {
```
?
```suggestion
		if ( ! class_exists( 'Automattic\\VIP\\Telemetry\\Tracks' ) ) {
```

chriszarate commented on inc/Analytics/EnvironmentConfig.php:
```diff
@@ -0,0 +1,81 @@
+		return 'local' === $this->tracks_core_props['vipgo_env'];
+	}
+
+	public function is_remote_data_blocks_plugin( string|null $plugin_path ): bool {
```
ideally these methods are `private` and unit tests target the public methods

mehmoodak replied:
Sorry didn't get you. We do require these methods as `public` so that they can be accessed in `TracksAnalytics` class.

chriszarate commented on inc/Analytics/TracksAnalytics.php:
```diff
@@ -0,0 +1,170 @@
+	 * @param int      $post_id Post ID.
+	 * @param \WP_Post $post Post object.
+	 */
+	public static function track_remote_data_blocks_usage( int $post_id, object $post ): void {
```
```suggestion
	public static function track_remote_data_blocks_usage( int $post_id, \WP_Post $post ): void {
```

mehmoodak replied:
The `object` type is used intentionally here, as using `WP_Post` directly in tests would require a tricky approach.

mehmoodak committed:
Merge branch 'trunk' into cafe-942/add-tracks-analytics

maxschmeling committed:
Merge branch 'trunk' into cafe-942/add-tracks-analytics

maxschmeling commented on tests/inc/Analytics/EnvironmentConfigTest.php:
```diff
@@ -0,0 +1,77 @@
+
+	public function testIsWpvipSiteReturnsTrue(): void {
+		$obj = new EnvironmentConfig();
+		set_private_property( EnvironmentConfig::class, $obj, 'tracks_core_props', [ 'hosting_provider' => 'wpvip' ] );
```
Do we absolutely need these private property methods?

Can we mock/stub out `get_tracks_core_properties` instead of setting `tracks_core_props` directly?

mehmoodak replied:
We're assigning `track_core_props` inside the `init` action hook, so using `set_private_property` seems okay given that we're not running tests within the WP environment.

maxschmeling reviewed (approved):
My only issue is with the private member stuff in the tests. If you think it's really necessary, I'm ok with it. But it seems like there could be a better way.

mehmoodak committed:
Merge branch 'trunk' into cafe-942/add-tracks-analytics
